src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -701,56 +694,84 @@ static ucs_status_t uct_tcp_sockcm_ep_client_init(uct_tcp_sockcm_ep_t *cep,
     return status;
 }
 
-static ucs_status_t uct_tcp_sockcm_ep_server_update_cm(uct_tcp_sockcm_ep_t *tcp_ep,
-                                                       uct_cm_h params_cm)
+static ucs_status_t uct_tcp_sockcm_ep_server_create(uct_tcp_sockcm_ep_t *tcp_ep,
+                                                    const uct_ep_params_t *params,
+                                                    uct_ep_h *ep_p)
 {
     uct_tcp_sockcm_t *tcp_sockcm        = uct_tcp_sockcm_ep_get_cm(tcp_ep);
-    uct_tcp_sockcm_t *params_tcp_sockcm = ucs_derived_of(params_cm, uct_tcp_sockcm_t);
-    ucs_async_context_t *async_ctx;
+    uct_tcp_sockcm_t *params_tcp_sockcm = ucs_derived_of(params->cm, uct_tcp_sockcm_t);
+    ucs_async_context_t *new_async_ctx;
     ucs_status_t status;
 
+    if (params->cm == NULL) {
+        ucs_error("cm cannot be NULL (ep=%p fd=%d)", tcp_ep, tcp_ep->fd);
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    UCS_ASYNC_BLOCK(tcp_sockcm->super.iface.worker->async);
+
     /* check if the server opened this ep, to the client, on a CM that is
      * different from the one it created its internal ep on earlier, when it
      * received the connection request from the client (the cm used by its listener) */
-    if (tcp_sockcm == params_tcp_sockcm) {
-        return UCS_OK;
+    if (tcp_sockcm != params_tcp_sockcm) {
```

Comment:
can we do this removal before taking lock?

Reply:
that would mean to remove something from the async handler while it's not under our control.
isn't better to have control over the thread and then work on it? 
is there a benefit to doing it?

Reply:
1. async API is thread safe, and we are not changing fields in cep itself
2. adding new fd is already done outside the lock
3. it can avoid potential deadlock when removing is waiting for handler to finish